File: src/main/java/leetcode/Problem560.java
package leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * https://leetcode.com/problems/subarray-sum-equals-k/
 */
public class Problem560 {
    public int subarraySum(int[] nums, int k) {
        int result = 0;
        int sum = 0;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            if (map.containsKey(sum - k)) {
                result += map.get(sum - k);
            }
            if (!map.containsKey(sum)) {
                map.put(sum, 1);
            } else {
                map.put(sum, map.get(sum) + 1);
            }
        }
        return result;
    }
}

File: src/main/rust/src/bin/problem560.rs
use std::collections::HashMap;

// https://leetcode.com/problems/subarray-sum-equals-k/description/
pub fn subarray_sum(nums: Vec<i32>, k: i32) -> i32 {
    let mut answer = 0;
    let mut map: HashMap<i32, i32> = HashMap::new();
    map.insert(0, 1);
    let mut sum = 0;
    for num in nums.iter() {
        sum += *num;
        if let Some(n) = map.get(&(sum - k)) {
            answer += *n;
        }
        *map.entry(sum).or_insert(0) += 1;
    }
    answer
}

fn main() {
    println!("{}", subarray_sum(vec![1, 1, 1], 2)); // 2
    println!("{}", subarray_sum(vec![1, 2, 3], 3)); // 2
    println!("{}", subarray_sum(vec![1, 2, 3, 1, 2], 3)); // 3
}
